keps/0022-diagnostics-bundle.md

```diff
@@ -0,0 +1,423 @@
+
+To configure diagnostics globally, this KEP introduces an optional top-level
+`diagnostics` key in operator.yaml.
+
+#### Diagnostics collection
+
+The following diagnostics will be implicitly collected without any configuration
+from the operator developer:
+
+- Logs for deployed pods related to the KUDO Instance
+- YAML for created resources, including both spec and status, related
+  to the KUDO instance
+- Output of `kubectl describe` for all deployed resources related to the KUDO
+  Instance
+- Current plan status, if one exists, or the KUDO Instance
+- Information about the KUDO instance's Operator and OperatorVersion
+- Logs for the KUDO controller manager
+- Describe for the KUDO controller manager resources
+- RBAC resources that are applicable to the KUDO controller manager
+- Current settings and version information for KUDO
+- Status of last preflight check run.
+
+Operator developer experience, then, focuses on customizing diagnostics
+information to gather information about the running application. The following
+forms are available, subject to change over time:
+
+- **Copy**: Copy a file out of a running pod. This is useful for non-stdout
+  logs, configuration files, and other artifacts generated by an application.
+  Higher level resources can also be used, which will copy the file on all pods
+  selected by that resource.
+- **Command**: Run a command on a running pod and copy the stdout. Higher level
+  resources can also be used, which will run the command on all pods selected by
+  that resource.
+- **Task**: Run a KUDO task and copy the stdout and other arbitrary files.
+- **HTTP**: Make an HTTP request from the KUDO controller manager to a named
+  service and port and copy the result of the request.
+
+While some of these are redundant (HTTP can be a command or job), the intent
+is to provide a high level experience where possible so that operator developers
+don't necessarily need to maintain a `curl` container as part of their
+application stack.
+
+Operator-defined diagnostics collection is defined in a new `diagnostics.bundle.resources`
+key in `operator.yaml`:
+
+```yaml
+diagnostics:
+  bundle:
+    resources:
+      - name: Zookeeper Configuration File
+        key: "zookeeper-configuration"
+        kind: Copy
+        spec:
+          path: /opt/zookeeper/server.properties
+          objectRef:
+            kind: StatefulSet # Runs on ALL pods in the statefulset
+            name: "{{ .InstanceName }}-zookeeper"
+      - name: DNS information for running pod
+        key: "dns-information"
+        kind: Command
+        spec:
+          command: # Can be string or array
+            - nslookup
+            - google.com
+          objectRef:
+            kind: Pod
+            name: "{{ .InstanceName }}-zookeeper-0"
+    filters:
+      - name: Authentication information
+        spec:
+          regex: "^host: %w+$"
+```
+
+This key is **OPTIONAL**. Default diagnostics collection will happen regardless
```

Comment:
Mostly because there's certain things around diagnostics that are fundamentally different from plans. For example, there's already at least one global-level diagnostics configuration setting. Also, there's an inherent "recovery" process that still needs to be determined. In Replicated and the DC/OS cases, the tools actually synchronously download the final tarball to the user's machine.

Reply:
I'm leaning more and more towards a client-side solution using the CLI. But just for the sake of the argument, you could easily model `diagnostics` as a plan with a series of tasks:
- each task would download a separate tarball (with logs or files) to your machine
- all tarball would end up in some `./diagnostic-dd.mm.yyyy-time/*` local folder
- you would automatically get all the workflow-engine benefits:
  - each step would run to the finish or retry on failure
  - you would see plan progress in the `Instance.Status`
  - etc.

Again, I'm in favor of the CLI-based solution so this is just a thought experiment ^^